binary_tree.c

```diff
@@ -0,0 +1,410 @@
+    }
+
+    self->r_node=new_node;
+}
+
+struct Node* tree_bfs_search(struct Node* self, int value) {
+    struct Node *n = NULL;
+    enQueue(self);
+
+    while((n = deQueue()) != NULL) {
+
+        if(n->data==value) {
+            return n;
+        }
+        if(n->l_node) {
+            enQueue(n->l_node);
+        }
+        if(n->r_node) {
+            enQueue(n->r_node);
+        }
+    }
+    return n;
+}
+
+struct Node* tree_dfs_search(struct Node* self, int value) {
+    /*
+    base case: value is found OR not
+    rec. step:
+        1. explore current node
+            if found: return value
+        2. if value is not in current node
+            a. recursive exploration of left subtree
+                if found: return value
+        3. if value is not in left subtree
+            a. recursive exploration of right subtree
+                if found: return value
+
+        if not found: return NULL
+    final state:
+        1. value is returned. algorithm is stopped.
+        2. NULL is returned. all nodes are visited
+    */
+
+   struct Node* n = NULL;
+
+    if(self->data == value) {
+        n = self;
+        return n;
+    }
+    if(self->l_node) {
+        n = tree_dfs_search(self->l_node, value);
+        if((n) && n->data == value) {
+            return n;
+        }
+    }
+    if(self->r_node) {
+        n = tree_dfs_search(self->r_node, value);
+        if((n) && n->data == value) {
+            return n;
+        }
+    }
+
+    return NULL;
+}
+
+void tree_postorder_traverse(struct Node* self) {
+    if (self->l_node) {
+        tree_postorder_traverse(self->l_node);
+    }
+    if (self->r_node) {
+        tree_postorder_traverse(self->r_node);
+    }
+
+    printf("%d\n", self->data);
+}
+
+void tree_preorder_traverse(struct Node* self) {
+    printf("%d\n", self->data);
+
+    if (self->l_node) {
+        tree_preorder_traverse(self->l_node);
+    }
+
+    if (self->r_node) {
+        tree_preorder_traverse(self->r_node);
+    }
+}
+
+void tree_inorder_traverse(struct Node* self) {
+    /*
+    base case:
+        all nodes in subtree are visited
+    rec. step:
+        if l_node
+            rec(l_node)
+        else
+            print data
+        print data
+        if r_node
+            rec(r_node)
+        else
+            print data
+    final state:
+        all nodes are visited
+    */
+
+    if (self->l_node) {
+        tree_inorder_traverse(self->l_node);
+    }
+
+    printf("%d\n", self->data);
+
+    if (self->r_node) {
+        tree_inorder_traverse(self->r_node);
+    }
+}
+
+void tree_delete_node(Tree* tree, struct Node* self) {
+    //if duplicate is encountered: delete first found node
+    struct Node* parent = tree_search_parent(tree, self);
+
+    if(self->l_node == NULL && self->r_node == NULL) {
+        free((void*) self);
+    }
+    else if(parent->l_node && parent->r_node == NULL) {
+        if(tree->root == self) {
+            tree->root = self->l_node;
```

Comment:
tree root shouldn'r be affected at all, imagine you have a tree:
```
    5
  3   8
2 4 7 9
     6
```
and you want to remove node 7. root shouldn't be affected at all. Later on, please write unit tests and tree_print

Reply:
I should have written comment that in this incorrect method until `else` I handled root with or without child nodes. In addition since the method is not recursive and has needless complexity